include/zephyr/bluetooth/classic/a2dp_codec_sbc.h

```diff
@@ -0,0 +1,98 @@
+/** @file
+ * @brief Advance Audio Distribution Profile - SBC Codec header.
+ */
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) 2015-2016 Intel Corporation
+ * Copyright (c) 2021 NXP
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_A2DP_CODEC_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_A2DP_CODEC_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Sampling Frequency */
+#define A2DP_SBC_SAMP_FREQ_16000 BIT(7)
+#define A2DP_SBC_SAMP_FREQ_32000 BIT(6)
+#define A2DP_SBC_SAMP_FREQ_44100 BIT(5)
+#define A2DP_SBC_SAMP_FREQ_48000 BIT(4)
+
+/* Channel Mode */
+#define A2DP_SBC_CH_MODE_MONO  BIT(3)
+#define A2DP_SBC_CH_MODE_DUAL  BIT(2)
+#define A2DP_SBC_CH_MODE_STREO BIT(1)
+#define A2DP_SBC_CH_MODE_JOINT BIT(0)
+
+/* Block Length */
+#define A2DP_SBC_BLK_LEN_4  BIT(7)
+#define A2DP_SBC_BLK_LEN_8  BIT(6)
+#define A2DP_SBC_BLK_LEN_12 BIT(5)
+#define A2DP_SBC_BLK_LEN_16 BIT(4)
+
+/* Subbands */
+#define A2DP_SBC_SUBBAND_4 BIT(3)
+#define A2DP_SBC_SUBBAND_8 BIT(2)
+
+/* Allocation Method */
+#define A2DP_SBC_ALLOC_MTHD_SNR      BIT(1)
+#define A2DP_SBC_ALLOC_MTHD_LOUDNESS BIT(0)
+
+#define BT_A2DP_SBC_SAMP_FREQ(cap)    ((cap->config[0] >> 4) & 0x0f)
+#define BT_A2DP_SBC_CHAN_MODE(cap)    ((cap->config[0]) & 0x0f)
+#define BT_A2DP_SBC_BLK_LEN(cap)      ((cap->config[1] >> 4) & 0x0f)
+#define BT_A2DP_SBC_SUB_BAND(cap)     ((cap->config[1] >> 2) & 0x03)
+#define BT_A2DP_SBC_ALLOC_MTHD(cap)   ((cap->config[1]) & 0x03)
+
+/** @brief SBC Codec */
+struct bt_a2dp_codec_sbc_params {
+	/** First two octets of configuration */
+	uint8_t config[2];
+	/** Minimum Bitpool Value */
+	uint8_t min_bitpool;
+	/** Maximum Bitpool Value */
+	uint8_t max_bitpool;
+} __packed;
+
+struct bt_a2dp_codec_sbc_media_packet_hdr {
+	uint8_t number_of_sbc_frames:4;
+	uint8_t rfa:1;
+	uint8_t L:1;
+	uint8_t S:1;
+	uint8_t F:1;
```

Comment:
This is not safe. The in-memory ordering of these bitfields will depend on the endianness of the architecture that you're building this for. It'd be better to just have a single `uint8_t` field and then use helper macros to access the individual bits.

Reply:
Thanks @jhedberg , I see Zephyr handle it by using CONFIG_LITTLE_ENDIAN. I can do the same way, is it OK? Because it is easy to use and write codes by using bit-fields.
```
/* RFC 4191, ch. 2.3 */
struct net_icmpv6_nd_opt_route_info {
	uint8_t prefix_len;
	struct {
#ifdef CONFIG_LITTLE_ENDIAN
		uint8_t reserved_2 :3;
		uint8_t prf        :2;
		uint8_t reserved_1 :3;
#else
		uint8_t reserved_1 :3;
		uint8_t prf        :2;
		uint8_t reserved_2 :3;
#endif
	} flags;
	uint32_t route_lifetime;
	/* Variable-length prefix field follows, can be 0, 8 or 16 bytes
	 * depending on the option length.
	 */
} __packed;
```

Reply:
@MarkWangChinese I'm fine either way. The needed duplication in struct definitions (or  internal subsets of a struct like in the above case) can get a bit messy and hard to follow, but in this case it seems it's fairy limited in scope. Btw, if possible, please double check that you got this correct by analyzing what actually gets generated by the compiler - I think we have at least some big endian architectures supported by Zephyr?

Reply:
Updated. I tried the gr716a_mini board with temporary bit-fields codes. The compiled assembly is expected/right.
```
struct test_struc {
	union {
		struct {
			uint8_t bit1:1;
			uint8_t bit2:2;
			uint8_t bit3:5
		} bit_test;
		uint8_t u8data;
	};
};

volatile struct test_struc test_data;

int main(void)
{
	test_data.bit_test.bit1 = 1;
        ......
}

	test_data.bit_test.bit1 = 1;
3100106c:	03 0c 00 01 	sethi  %hi(0x30000400), %g1
31001070:	c4 08 61 d8 	ldub  [ %g1 + 0x1d8 ], %g2	! 300005d8 <test_data>
31001074:	84 10 bf 80 	or  %g2, -128, %g2
31001078:	c4 28 61 d8 	stb  %g2, [ %g1 + 0x1d8 ]
```